### PR TITLE
Updated Exodus decayer

### PR DESCRIPTION
**Pull request by otonvd**
Changes as well in alipythia to allow new decays to be handled by Exodus

### CODE REVIEW

File: EVGEN/AliDecayerExodus.cxx [hunk 1]
```diff
@@ -51,11 +51,13 @@ AliDecayerExodus::AliDecayerExodus():
     fEPMassPion(0),
     fEPMassEta(0),
     fEPMassEtaPrime(0),
+fEPMassEtaPrime_toOmega(0),
```
Comment (dberzano) [inline EVGEN/AliDecayerExodus.cxx]: Fix indentation

File: EVGEN/AliDecayerExodus.cxx [hunk 2]
```diff
@@ -51,11 +51,13 @@ AliDecayerExodus::AliDecayerExodus():
...
     fEPMassRho(0),
     fEPMassOmega(0),
     fEPMassOmegaDalitz(0),
     fEPMassPhi(0),
     fEPMassPhiDalitz(0),
+fEPMassPhiDalitz_toPi0(0),
```
Comment (dberzano) [inline EVGEN/AliDecayerExodus.cxx]: Same here

File: EVGEN/AliDecayerExodus.cxx [hunk 3]
```diff
@@ -113,54 +115,68 @@ void AliDecayerExodus::Init()
     maxeta      = etamass - omassgamma;
     maxomega    = omegamass - pionmass;
     maxetaprime = etaprimemass - omassgamma;
+maxetaprime_toOmega = etaprimemass - omegamass;
```
Comment (dberzano) [inline EVGEN/AliDecayerExodus.cxx]: Indentation again

File: EVGEN/AliDecayerExodus.cxx (hunk 1 above)
Comment (otonvd) [reply EVGEN/AliDecayerExodus.cxx]: Hi Dario, should  I close the pull request and make a new one?

Comment (dberzano) [reply EVGEN/AliDecayerExodus.cxx]: If you keep pushing to your current branch, the PR will be updated. See [here](http://alisw.github.io/git-tutorial/#amend-an-open-pull-request).

**Comment by rbertens** [comment]
+test

**Comment by ktf** [comment]
@dberzano is there anything stuck for the PR checking? This seems to be stalling...

**Comment by ktf** [comment]
+test

**Comment by otonvd** [comment]
Dear all, since this is one of my first commits using github, I am now a bit lost. What's the status of this commit? Is there any action required from my side?

**Comment by dberzano** [comment]
+test

As the last comment says, one of the mentioned persons must approve the commit. Approval is required every time you modify the PR (not just when you open it). Some expert on the given matter should give the ultimate approval if this PR compiles (I have approved test runs without automatic merging).

**Comment by dberzano** [comment]
I believe there is something wrong with the AliRoot tester. Let me restart it. The error will clear out automatically, ignore it for now.

**Comment by otonvd** [comment]
Unfortunately I cannot see the log...

**Comment by dberzano** [comment]
@otonvd the infrastructure has temporary problems that do not depend on us (but CERN IT). The message is misleading, but please ignore it. The PR will become green as soon as the problem is fixed.

**Comment by otonvd** [comment]
I see that I create a pull request everytime I force push (the only way I can actually push). I don't know how to solve it!

**Comment by rbertens** [comment]
Hi @otonvd , would be good if @dberzano  could confirm, but I think this happens because as long as this current PR is open, every time you push, the changes you've made are added as an 'addendum' to this open PR.

**Comment by miweberSMI** [comment]
Hi all,

what is the status of this PR? I thought it shoudl be approved by now?

Thanks and cheers,
Michael

**Comment by dberzano** [comment]
> I see that I create a pull request everytime I force push (the only way I can actually push). I don't know how to solve it!

There's nothing to solve there. It is on purpose, @rbertens is right. When you force-push, you change hash. New hashes are new commits (we identify commits by hash). When a new hash is pushed, approval has to be requested for that new hash. We cannot reuse the approval for another hash, the reason being: imagine if you get the approval for some content, and then you change it, and we keep the former approval...

Besides: approval from PWGMM is required before I approve (I am admin, if I approve my part, it gets merged right away, so PWGMM must `+1` first). This is not a formality, PWGMM should really check what you have done and see if it's OK or not, so please wait for @rbertens or @qgp first.

**Comment by dberzano** [comment]
+test

**Comment by rbertens** [comment]
Hi @miweberSMI , @otonvd , all,

The changes to EVGEN/* are fine from my side, but since this PR also includes some changes to PYTHIA6 (which are also outside of what I can '+1') it would be good to have the input from @qgp . 

More generally, as this changes some fundamental code used by multiple generators, it would be very useful if we could have some additional information, e.g.: was this code validated? is there a JIRA ticket connected? was it discussed in a broader scope? 

Thanks and apologies for the delay,

Redmer

**Comment by ktf** [comment]
@dberzano any reason for not merging this?

**Comment by dberzano** [comment]
Yes, waiting for @rbertens or @qgp approval.

**Comment by rbertens** [comment]
HI all, I gave a comment already quite a wihle ago:

> The changes to EVGEN/* are fine from my side, but since this PR also includes some changes to PYTHIA6 (which are also outside of what I can '+1') it would be good to have the input from @qgp .

> More generally, as this changes some fundamental code used by multiple generators, it would be very useful if we could have some additional information, e.g.: was this code validated? is there a JIRA ticket connected? was it discussed in a broader scope?

> Thanks and apologies for the delay,

> Redmer

**Comment by miweberSMI** [comment]
Hello @dberzano, all,

I am wondering, what happened to this pull request?

thanks and cheers,
Michael

**Comment by dberzano** [comment]
@miweberSMI we are waiting for a feedback from @qgp, who's been on vacation for quite some time. @rbertens said that it would be good to have a feedback from him before approving.

**Comment by miweberSMI** [comment]
OK, thanks for clarification.

**Comment by miweberSMI** [comment]
Hi @qgp,

are there any opinion from your side on this pull reuqest? We would like to have it for another porting request for LHC15o productions...

Thanks and cheers,
Michael

File: EVGEN/AliDecayerExodus.cxx [hunk 4]
```diff
@@ -346,6 +384,7 @@ Double_t AliDecayerExodus::Lorentz(Float_t mass, Double_t vmass, Double_t vwidth
 
 }
 
+//void AliDecayerExodus::Decay(Int_t idpart, TLorentzVector* pparent, Int_t idpartner)
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: should be removed

File: EVGEN/AliDecayerExodus.cxx [hunk 5]
```diff
@@ -804,7 +860,8 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if((dp3[0]->GetPdgCode() != 22) && (TMath::Abs(dp3[1]->GetPdgCode()) != 11))   continue;
 
   TLorentzVector Pizero(part->Px(), part->Py(), part->Pz(), part->Energy());
-  Decay(111, &Pizero);
+  //  Decay(111, &Pizero, dp3[0]->GetPdgCode());
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 6]
```diff
@@ -826,7 +883,8 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if((dp3[0]->GetPdgCode() != 22) && ((TMath::Abs(dp3[1]->GetPdgCode()) != 11)))   continue;
 
   TLorentzVector Eta(part->Px(), part->Py(), part->Pz(), part->Energy());
-  Decay(221, &Eta);
+  //Decay(221, &Eta, dp3[0]->GetPdgCode());
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 7]
```diff
@@ -847,6 +905,7 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if((dp2[0]->GetPdgCode() != 11) && ((TMath::Abs(dp2[1]->GetPdgCode()) != 11)))   continue;
 
   TLorentzVector Rho(part->Px(), part->Py(), part->Pz(), part->Energy());
+  //Decay(113, &Rho, 0);
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 8]
```diff
@@ -868,7 +927,8 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if( dp3[0]->GetPdgCode() != 111 && (TMath::Abs(dp3[1]->GetPdgCode()) != 11)) continue;
 
   TLorentzVector Omegadalitz(part->Px(), part->Py(), part->Pz(), part->Energy());
-  Decay(223, &Omegadalitz);
+  //Decay(223, &Omegadalitz, dp3[0]->GetPdgCode());
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 9]
```diff
@@ -897,10 +958,11 @@ void AliDecayerExodus::Decay(TClonesArray *array)
 
   for (j = 0; j < 3; j++) dp3[j] = (TParticle*) (array->At(fd3+j));
 
-  if((dp3[0]->GetPdgCode() != 22) && ((TMath::Abs(dp3[1]->GetPdgCode()) != 11)))   continue;
+  if((dp3[0]->GetPdgCode() != 22 && dp3[0]->GetPdgCode() != 223) && ((TMath::Abs(dp3[1]->GetPdgCode()) != 11)))   continue;
 
   TLorentzVector Etaprime(part->Px(), part->Py(), part->Pz(), part->Energy());
-  Decay(331, &Etaprime);
+  //Decay(331, &Etaprime, dp3[0]->GetPdgCode());
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 10]
```diff
@@ -918,7 +980,8 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if( dp3[0]->GetPdgCode() != 221 && (TMath::Abs(dp3[1]->GetPdgCode()) != 11)) continue;
 
   TLorentzVector Phidalitz(part->Px(), part->Py(), part->Pz(), part->Energy());
-  Decay(333, &Phidalitz);
+  //Decay(333, &Phidalitz, dp3[0]->GetPdgCode());
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 11]
```diff
@@ -927,6 +990,7 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if( dp2[0]->GetPdgCode() != 11 && (TMath::Abs(dp2[1]->GetPdgCode()) != 11))   continue;
 
   TLorentzVector Phi(part->Px(), part->Py(), part->Pz(), part->Energy());
+  //Decay(333, &Phi, 0);
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.cxx [hunk 12]
```diff
@@ -949,6 +1013,7 @@ void AliDecayerExodus::Decay(TClonesArray *array)
   if((dp2[0]->GetPdgCode() != 11) && ((TMath::Abs(dp2[1]->GetPdgCode()) != 11)))   continue;
 
   TLorentzVector JPsi(part->Px(), part->Py(), part->Pz(), part->Energy());
+  //Decay(443, &JPsi, 0);
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.cxx]: remove

File: EVGEN/AliDecayerExodus.h [hunk 13]
```diff
@@ -35,6 +35,7 @@ class AliDecayerExodus : public AliDecayer
  public:
     AliDecayerExodus();
     virtual void    Init();
+  //  virtual void    Decay(Int_t idpart,TLorentzVector* pparent, Int_t idpartner);
```
Comment (qgp) [inline EVGEN/AliDecayerExodus.h]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 14]
```diff
@@ -1588,7 +1588,8 @@ void AliPythia::PizeroDalitz()
   continue;
   TLorentzVector pizero(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
-  fExodus.Decay(pdg, &pizero);
+  //fExodus.Decay(pdg, &pizero, fPyjets->K[1][fd]);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 15]
```diff
@@ -1613,7 +1614,8 @@ void AliPythia::EtaDalitz()
   continue;
   TLorentzVector eta(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
-  fExodus.Decay(pdg, &eta);
+  //fExodus.Decay(pdg, &eta,fPyjets->K[1][fd] );
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 16]
```diff
@@ -1638,6 +1640,7 @@ void AliPythia::RhoDirect()
   continue;
   TLorentzVector rho(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
+  //fExodus.Decay(pdg, &rho, 0);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 17]
```diff
@@ -1663,7 +1666,8 @@ void AliPythia::OmegaDalitz()
   continue;
   TLorentzVector omegadalitz(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
-  fExodus.Decay(pdg, &omegadalitz);
+  //fExodus.Decay(pdg, &omegadalitz, fPyjets->K[1][fd]);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 18]
```diff
@@ -1688,6 +1692,7 @@ void AliPythia::OmegaDirect()
   continue;
   TLorentzVector omegadirect(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
+  //fExodus.Decay(pdg, &omegadirect, 0);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 19]
```diff
@@ -1709,11 +1714,12 @@ void AliPythia::EtaprimeDalitz()
   Int_t ld = fPyjets->K[4][i] - 1;
   if (fd < 0)                                        continue;
   if ((ld - fd) != 2)                                continue;
-  if ((fPyjets->K[1][fd] != 22) || (TMath::Abs(fPyjets->K[1][fd+1]) != 11))
+  if ((fPyjets->K[1][fd] != 22 && fPyjets->K[1][fd] != 223) || (TMath::Abs(fPyjets->K[1][fd+1]) != 11))
   continue;
   TLorentzVector etaprime(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
-  fExodus.Decay(pdg, &etaprime);
+  //fExodus.Decay(pdg, &etaprime, fPyjets->K[1][fd]);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 20]
```diff
@@ -1734,11 +1740,12 @@ void AliPythia::PhiDalitz()
   Int_t ld = fPyjets->K[4][i] - 1;
   if (fd < 0)                                        continue;
   if ((ld - fd) != 2)                                continue;
-  if ((fPyjets->K[1][fd] != 221) || (TMath::Abs(fPyjets->K[1][fd+1]) != 11))
+  if ((fPyjets->K[1][fd] != 221 && fPyjets->K[1][fd] != 111 ) || (TMath::Abs(fPyjets->K[1][fd+1]) != 11))
   continue;
   TLorentzVector phidalitz(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
-  fExodus.Decay(pdg, &phidalitz);
+  //fExodus.Decay(pdg, &phidalitz, fPyjets->K[1][fd]);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 21]
```diff
@@ -1763,6 +1770,7 @@ void AliPythia::PhiDirect()
   continue;
   TLorentzVector phi(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
+  //fExodus.Decay(pdg, &phi, 0);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

File: PYTHIA6/AliPythia6/AliPythia.cxx [hunk 22]
```diff
@@ -1787,6 +1795,7 @@ void AliPythia::JPsiDirect()
   continue;
   TLorentzVector jpsi(fPyjets->P[0][i], fPyjets->P[1][i], fPyjets->P[2][i], fPyjets->P[3][i]);
   Int_t pdg = TMath::Abs(fPyjets->K[1][i]);
+  //fExodus.Decay(pdg, &jpsi, 0);
```
Comment (qgp) [inline PYTHIA6/AliPythia6/AliPythia.cxx]: remove

**Comment by otonvd** [comment]
Dear all,
I cleaned the code as suggested. With respect to a more detailed description of this code and its possible implications for any pythia simulation:

- All changes in the code are affecting exclusively LMee simulations with the Exodus decayer.
- For the Pythia part of the code, still only affecting the Exodus decayer, in particular what is done here is to accommodate the code for handling two new decays added in the decay table to dielectrons (namely eta' -> omega e+ e- and phi -> pi0 e+ e-).

Now one question regarding the pull request. Since the time when it was first proposed I changed my AliRoot installations, folders, etc. Now I have an updated AliRoot and the changes with respect to the master are still only this 5 files that I want to pull, but I don't know if I do it right now it will be part of this old pull request or it will be done as a new one. Any help on this?

Cheers
oton

**Comment by otonvd** [comment]
Not having received any suggestion on how to do it, I cancel this pull request and create a new one with the same files.

**Comment by miweberSMI** [comment]
Hi @otonvd,  

can you mention the new pull request here?

Thanks 
Michael

**Comment by otonvd** [comment]
New pull request is: #549 
https://github.com/alisw/AliRoot/pull/549

**Pull request closed without merging**